File: demo/java/src/main/java/io/github/cosinekitty/astronomy/demo/RiseSetCulm.java
package io.github.cosinekitty.astronomy.demo;

import java.util.ArrayList;
import java.util.Collections;

import io.github.cosinekitty.astronomy.*;

public class RiseSetCulm {
    /**
     * Display rise/set/culmination times for the Sun and Moon.
     *
     * @param observer
     * The geographic location of the observer.
     *
     * @param startTime
     * The date and time after which to search for the first
     * occurrences of rise/set/culmination times of the Sun and Moon.
     */
    public static int run(Observer observer, Time startTime) {
        // At extreme latitudes (close to the poles), there can be long
        // periods of time where the Sun/Moon never rise/set.
        // Allow the search to proceed for 2 days, which will usually find events
        // for most populated latitudes.
        // In other applications (e.g. a daily calendar),
        // the developer may wish to use a 1-day search limit.
        // Note that the Moon rises/sets about 1 hour later each day, so it is
        // common for the Moon to not have a rise or set event in a given day.
        final double dayLimit = 2.0;

        var eventList = new ArrayList<AstroEvent>();

        // Rise/set times may or may not occur within any finite search window.
        maybeAdd(eventList, "sunrise",  Astronomy.searchRiseSet(Body.Sun,  observer, Direction.Rise, startTime, dayLimit));
        maybeAdd(eventList, "sunset",   Astronomy.searchRiseSet(Body.Sun,  observer, Direction.Set,  startTime, dayLimit));
        maybeAdd(eventList, "moonrise", Astronomy.searchRiseSet(Body.Moon, observer, Direction.Rise, startTime, dayLimit));
        maybeAdd(eventList, "moonset",  Astronomy.searchRiseSet(Body.Moon, observer, Direction.Set,  startTime, dayLimit));

        // Culmination times can always be found regardless of latitude.
        eventList.add(new AstroEvent("sunculm",  Astronomy.searchHourAngle(Body.Sun,  observer, 0.0, startTime, +1)));
        eventList.add(new AstroEvent("moonculm", Astronomy.searchHourAngle(Body.Moon, observer, 0.0, startTime, +1)));

        // Sort the list chronologically.
        Collections.sort(eventList);

        // Print the events.
        for (AstroEvent e : eventList) {
            e.display();
        }

        return 0;
    }

    private static void maybeAdd(ArrayList<AstroEvent> eventList, String name, Time time) {
        if (time != null) {
            eventList.add(new AstroEvent(name, time));
        }
    }

    private static class AstroEvent implements Comparable<AstroEvent> {
        public final String name;
        public final Time time;
        public final double altitude;

        public AstroEvent(String name, Time time) {
            this.name = name;
            this.time = time;
            this.altitude = Double.NaN;     // this event has no altitude data associated with it
        }

        public AstroEvent(String name, HourAngleInfo hourAngleInfo) {
            this.name = name;
            this.time = hourAngleInfo.getTime();
            this.altitude = hourAngleInfo.getHor().getAltitude();
        }

        @Override
        public int compareTo(AstroEvent other) {
            return this.time.compareTo(other.time);
        }

        public void display() {
            System.out.print(String.format("%-8s : %s", name, time));
            if (Double.isFinite(altitude)) {
                System.out.print(String.format("   altitude = %5.2f", altitude));
            }
            System.out.println();
        }
    }
}

File: demo/kotlin/src/main/kotlin/RiseSetCulm.kt
import kotlin.collections.sorted
import io.github.cosinekitty.astronomy.*

/**
 * Display rise/set/culmination times for the Sun and Moon.
 *
 * @param observer
 * The geographic location of the observer.
 *
 * @param startTime
 * The date and time after which to search for the first
 * occurrences of rise/set/culmination times of the Sun and Moon.
 */
internal fun `Rise Set Culmination demo`(observer: Observer, startTime: Time): Int {
    // At extreme latitudes (close to the poles), there can be long
    // periods of time where the Sun/Moon never rise/set.
    // Allow the search to proceed for 2 days, which will usually find events
    // for most populated latitudes.
    // In other applications (e.g. a daily calendar),
    // the developer may wish to use a 1-day search limit.
    // Note that the Moon rises/sets about 1 hour later each day, so it is
    // common for the Moon to not have a rise or set event in a given day.
    val dayLimit = 2.0
    val eventList = listOfNotNull(
        // Rise/set times may or may not occur within any finite search window.
        maybe("sunrise",  searchRiseSet(Body.Sun,  observer, Direction.Rise, startTime, dayLimit)),
        maybe("sunset",   searchRiseSet(Body.Sun,  observer, Direction.Set,  startTime, dayLimit)),
        maybe("moonrise", searchRiseSet(Body.Moon, observer, Direction.Rise, startTime, dayLimit)),
        maybe("moonset",  searchRiseSet(Body.Moon, observer, Direction.Set,  startTime, dayLimit)),

        // Culmination times can always be found regardless of latitude.
        AstroEvent("sunculm",  searchHourAngle(Body.Sun,  observer, 0.0, startTime)),
        AstroEvent("moonculm", searchHourAngle(Body.Moon, observer, 0.0, startTime)),
    )

    // Print the list in sorted order.
    eventList.sorted().forEach(AstroEvent::display)

    return 0
}


private fun maybe(name: String, time: Time?): AstroEvent? = time?.let { AstroEvent(name, it) }


private class AstroEvent(
    val name: String,
    val time: Time,
    val altitude: Double? = null
): Comparable<AstroEvent> {
    // Allow a list of AstroEvent to be sorted chronologically.
    override operator fun compareTo(other: AstroEvent): Int = this.time.compareTo(other.time)

    constructor(name: String, hourAngleInfo: HourAngleInfo)
        : this(name, hourAngleInfo.time, hourAngleInfo.hor.altitude)

    fun display() {
        print("%-8s : %s".format(name, time))
        if (altitude != null) {
            print("   altitude = %5.2f".format(altitude))
        }
        println()
    }
}
